src/neon/stage1.rs

```diff
@@ -0,0 +1,603 @@
+
+// NEON-SPECIFIC
+
+macro_rules! bit_mask {
+    () => {
+        uint8x16_t::new(
+            0x01, 0x02, 0x4, 0x8, 0x10, 0x20, 0x40, 0x80,
+            0x01, 0x02, 0x4, 0x8, 0x10, 0x20, 0x40, 0x80
+        )
+    };
+}
+
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn neon_movemask(input: uint8x16_t) -> u16 {
+    let minput: uint8x16_t = vandq_u8(input, bit_mask!());
+    let tmp: uint8x16_t = vpaddq_u8(minput, minput);
+    let tmp = vpaddq_u8(tmp, tmp);
+    let tmp = vpaddq_u8(tmp, tmp);
+
+    vgetq_lane_u16(vreinterpretq_u16_u8(tmp), 0)
+}
+
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn neon_movemask_bulk(p0: uint8x16_t, p1: uint8x16_t, p2: uint8x16_t, p3: uint8x16_t) -> u64 {
+    let bit_mask = bit_mask!();
+
+    let t0 = vandq_u8(p0, bit_mask);
+    let t1 = vandq_u8(p1, bit_mask);
+    let t2 = vandq_u8(p2, bit_mask);
+    let t3 = vandq_u8(p3, bit_mask);
+    let sum0 = vpaddq_u8(t0, t1);
+    let sum1 = vpaddq_u8(t2, t3);
+    let sum0 = vpaddq_u8(sum0, sum1);
+    let sum0 = vpaddq_u8(sum0, sum0);
+
+    vgetq_lane_u64(vreinterpretq_u64_u8(sum0), 0)
+}
+
+// /NEON-SPECIFIC
+
+pub const SIMDJSON_PADDING: usize = mem::size_of::<uint8x16_t>() * 4;
+
+unsafe fn compute_quote_mask(quote_bits: u64) -> u64 {
+    vgetq_lane_u64(
+        vreinterpretq_u64_u8(
+            mem::transmute(
+                vmull_p64(
+                    -1,
+                    quote_bits as i64)
+            )
+        ),
+        0
+    )
+}
+
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn check_ascii(si: &SimdInput) -> bool {
+    let highbit: uint8x16_t = vdupq_n_u8(0x80);
+    let t0: uint8x16_t = vorrq_u8(si.v0, si.v1);
+    let t1: uint8x16_t = vorrq_u8(si.v2, si.v3);
+    let t3: uint8x16_t = vorrq_u8(t0, t1);
+    let t4: uint8x16_t = vandq_u8(t3, highbit);
+
+    let v64: uint64x2_t = vreinterpretq_u64_u8(t4);
+    let v32: uint32x2_t = vqmovn_u64(v64);
+    let result: uint64x1_t = vreinterpret_u64_u32(v32);
+
+    vget_lane_u64(result, 0) == 0
+}
+
+#[derive(Debug)]
+struct SimdInput {
+    v0: uint8x16_t,
+    v1: uint8x16_t,
+    v2: uint8x16_t,
+    v3: uint8x16_t,
+}
+
+fn fill_input(ptr: &[u8]) -> SimdInput {
+    unsafe {
+        #[allow(clippy::cast_ptr_alignment)]
+        SimdInput {
+            v0: vld1q_u8(ptr.as_ptr() as *const u8),
+            v1: vld1q_u8(ptr.as_ptr().add(16) as *const u8),
+            v2: vld1q_u8(ptr.as_ptr().add(32) as *const u8),
+            v3: vld1q_u8(ptr.as_ptr().add(48) as *const u8),
+        }
+    }
+}
+
+struct Utf8CheckingState {
+    has_error: int8x16_t,
+    previous: ProcessedUtfBytes,
+}
+
+impl Default for Utf8CheckingState {
+    #[cfg_attr(not(feature = "no-inline"), inline)]
+    fn default() -> Self {
+        Utf8CheckingState {
+            has_error: vdupq_n_s8(0),
+            previous: ProcessedUtfBytes::default(),
+        }
+    }
+}
+
+#[inline]
+fn is_utf8_status_ok(has_error: int8x16_t) -> bool {
```

Comment:
I'm a bit confused about this. 'AND' the data with itself, which has no effect and then turn it into a 128 bit type and then into a 16 bit type and then compare it to zero.

What is it we actually trying to do? Check that everything i there is zero?

Reply:
yeah that was confusing to me too, it's widening over the whole bit width... return 1 if AND(a, b) is all zero bits

Reply:
So `vandq_s16` is bitewise and, not logical [1] so `vandq_s16(has_error, has_error) == has_error`, then we pull it out as a single 128b integer and extract only the last 16 (so in the case the last hightest), ignoring 1-3 all others.

I'm worried that this passes our test cases but is logically incorrect, at least all I can come up with screams 'this is wrong'.

Partially because A && A seems obsolute, and we're down-casting and throwing away a lot of bits.


[1] https://developer.arm.com/architectures/instruction-sets/simd-isas/neon/intrinsics?search=vandq_s16

Reply:
So what also worries me is that if I remove the and in there all tests pass, so either it really has no effect or the tests do not cover any case where it does.

And even with the  and back in there I could construct a test case that panics this. (fail34 with the bad utf8 moved up forward)

Reply:
```diff
diff --git a/data/fail/fail34.json b/data/fail/fail34.json
index e82c700..0445278 100644
--- a/data/fail/fail34.json
+++ b/data/fail/fail34.json
@@ -1 +1 @@
-["this string contains bad UTF-8 <80>"]
+[" <80>"]
diff --git a/src/neon/stage1.rs b/src/neon/stage1.rs
index b23f059..65e2535 100644
--- a/src/neon/stage1.rs
+++ b/src/neon/stage1.rs
@@ -114,13 +114,10 @@ impl Default for Utf8CheckingState {
 fn is_utf8_status_ok(has_error: int8x16_t) -> bool {
     unsafe {
         let utf8_error_bits: u128 = mem::transmute(
-            vandq_s16(
-                mem::transmute(has_error),
-                mem::transmute(has_error)
-            )
+            has_error
         );

-        utf8_error_bits as u16 == 0
+        utf8_error_bits == 0
     }
 }
```

This passes correctly for.

I'll add some fail cases that cover 'bad' utf8 in different places of the input, this just slipped through since we only had it in one location and that happend  to fall into the 16 bit.

Reply:
@Licenser great catch!